internal/ecies_pkcs11.go

```diff
@@ -216,38 +146,44 @@ func EciesDecrypt(prv PrivateKey, c, s1, s2 []byte) (m []byte, err error) {
 	case 2, 3, 4:
 		rLen = (pub.Curve.Params().BitSize + 7) / 4
 		if len(c) < (rLen + hLen + 1) {
-			return nil, ErrInvalidMessage
+			return nil, ecies.ErrInvalidMessage
 		}
 	default:
-		return nil, ErrInvalidPublicKey
+		return nil, ecies.ErrInvalidPublicKey
 	}
 
 	mStart = rLen
 	mEnd = len(c) - hLen
 
-	R := new(ecdsa.PublicKey)
+	R := new(ecies.PublicKey)
 	R.Curve = pub.Curve
 	R.X, R.Y = elliptic.Unmarshal(R.Curve, c[:rLen])
 	if R.X == nil {
-		return nil, ErrInvalidPublicKey
+		return nil, ecies.ErrInvalidPublicKey
+	}
+	if !R.Curve.IsOnCurve(R.X, R.Y) {
+		err = ecies.ErrInvalidCurve
+		return
 	}
 
 	z, err := prv.GenerateShared(R, params.KeyLen, params.KeyLen)
 	if err != nil {
 		return nil, err
 	}
+
 	Ke, Km := deriveKeys(hash, z, s1, params.KeyLen)
 	if Ke == nil || Km == nil {
-		return nil, ErrInvalidPublicKey
+		return nil, ecies.ErrInvalidPublicKey
 	}
 
 	d := messageTag(params.Hash, Km, c[mStart:mEnd], s2)
 	if d == nil {
-		return nil, ErrInvalidMessage
+		return nil, ecies.ErrInvalidMessage
 	}
 	if subtle.ConstantTimeCompare(c[mEnd:], d) != 1 {
-		return nil, ErrInvalidMessage
+		return nil, ecies.ErrInvalidMessage
 	}
 
-	return symDecrypt(params, Ke, c[mStart:mEnd])
+	// rand is unused below, but mandatory per some legacy interface
+	return symDecrypt(nil, params, Ke, c[mStart:mEnd])
```

Comment:
It is not even used by the implementation in ecies lib.

Reply:
yep... it looks like some rudiment; so much was my surprise when I saw that argument.